src/arch/loongarch64/core/la64_nap.c

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2022, Loongson Technology Corporation Limited. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ *
+ * You can also choose to distribute this program under the terms of
+ * the Unmodified Binary Distribution Licence (as given in the file
+ * COPYING.UBDL), provided that you have satisfied its requirements.
+ */
+
+FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );
+
+#include <ipxe/nap.h>
+#include <ipxe/efi/efi.h>
+
+/** @file
+ *
+ * iPXE CPU sleeping API for EFI
+ *
+ */
+
+/**
+ * Sleep until next interrupt
+ *
+ */
+static void la64_cpu_nap ( void ) {
+	/*
+	 * I can't find any EFI API that allows us to put the CPU to
+	 * sleep.  The CpuSleep() function is defined in CpuLib.h, but
+	 * isn't part of any exposed protocol so we have no way to
+	 * call it.
+	 *
+	 * The EFI shell doesn't seem to bother sleeping the CPU; it
+	 * just sits there idly burning power.
+	 *
+	 */
+	__asm__ __volatile__ ( "idle 1" );
```

Comment:
[IDLE](https://loongson.github.io/LoongArch-Documentation/LoongArch-Vol1-EN.html#_idle) is a privileged instruction, but it works fine on both physical and virtual machines.
So, can I use it here?

Reply:
> [IDLE](https://loongson.github.io/LoongArch-Documentation/LoongArch-Vol1-EN.html#_idle) is a privileged instruction, but it works fine on both physical and virtual machines. So, can I use it here?

Yes, but only if the code is renamed to make it clear that it's EFI-specific.

Since it's a privileged instruction then it needs to be attempted only when we are running in PLV0 (either real PLV0 on physical hardware, or virtual PLV0 inside a VM or emulator).

An EFI binary will execute at (real or virtual) PLV0 and so using IDLE is fine there.

A Linux userspace binary (even if running as root) will execute at PLV3 and so using IDLE will presumably crash the process.  You won't have picked this up from running the self-test suite alone, since the test suite never idles.  I've added a self-test to cover this just now in commit https://github.com/ipxe/ipxe/commit/66a2ff442

You won't actually hit the crash under Linux even with the new test, because you only pull in this nap code using IDLE for the LoongArch64 EFI builds.  But since this is the case, we should name this code appropriately: it's not generic LoongArch64 napping code, it's EFI-specific LoongArch64 napping code.

Reply:
With the last commit, the self-test pass.